### PR TITLE
Optimizations on execution metrics api

### PR DESCRIPTION
**Pull request by ahormazabal**
- Made refactorings on the query API to make it more reusable
- Made optimizations on metrics calculation to use database projections.
- Removed status counting from statistics to reduce the impact of the solution. Metrics for each status can still be obtained by calling the API using the status filter.

### CODE REVIEW

**Comment by ahormazabal** [comment]
Documentation update: https://github.com/rundeck/docs/pull/225

File: rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy [hunk 1]
```diff
@@ -143,4 +143,287 @@ class ExecutionQuery extends ScheduledExecutionQuery implements Validateable{
...
+        }
+      }
+      def idlist = query.jobIdListFilter?.collect(convertids)
+      def xidlist = query.excludeJobIdListFilter?.collect(convertids)
+
+      def hasJobFilters = jobqueryfilters.any { query[it] }
+      if (hasJobFilters && query.adhoc) {
+        flash.errorCode = "api.error.parameter.error"
+        flash.errorArgs = [message(code: 'api.executions.jobfilter.adhoc.conflict')]
+        return chain(controller: 'api', action: 'renderError')
```
Comment (gschueler) [inline rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy]: does this work from within this method? handling errors/response should probably be moved out of here

Comment (gschueler) [reply rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy]: e.g. just throw exception here and catch it in the controller

**Comment by gschueler** [comment]
Aside from the error response issue, looks good

File: rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy (hunk 1 above)
Comment (ahormazabal) [reply rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy]: improved handling of error messages

**Comment by ahormazabal** [comment]
Documentation Update: https://github.com/rundeck/docs/pull/249

**Pull request merged**